marconi/src/Marconi/Index/New/ScriptTx.hs

```diff
@@ -0,0 +1,418 @@
+
+deriving anyclass instance SQL.FromRow ScriptTxRow
+
+-- * Indexer
+
+type Query = StorableQuery ScriptTxHandle
+type Result = StorableResult ScriptTxHandle
+
+toUpdate
+  :: forall era . C.IsCardanoEra era
+  => [C.Tx era]
+  -> SlotNo
+  -> StorableEvent ScriptTxHandle
+toUpdate txs = ScriptTxEvent txScripts'
+  where
+    txScripts' = map (\tx -> (TxCbor $ C.serialiseToCBOR tx, getTxScripts tx)) txs
+
+getTxBodyScripts :: forall era . C.TxBody era -> [StorableQuery ScriptTxHandle]
+getTxBodyScripts body = let
+    hashesMaybe :: [Maybe C.ScriptHash]
+    hashesMaybe = case body of
+      Shelley.ShelleyTxBody shelleyBasedEra _ scripts _ _ _ ->
+        flip map scripts $ \script ->
+          case fromShelleyBasedScript shelleyBasedEra script of
+            Shelley.ScriptInEra _ script' -> Just $ C.hashScript script'
+      _ -> [] -- Byron transactions have no scripts
+    hashes = catMaybes hashesMaybe :: [Shelley.ScriptHash]
+  in map ScriptTxAddress hashes
+
+getTxScripts :: forall era . C.Tx era -> [StorableQuery ScriptTxHandle]
+getTxScripts (C.Tx txBody _ws) = getTxBodyScripts txBody
+
+{- Now that all connected data types have been defined, we go on to implement some
+   of the type classes required for information storage and retreival. -}
+
+instance Buffered ScriptTxHandle where
+  {- The data is buffered in memory. When the memory buffer is filled, we need to store
+     it on disk. -}
+  persistToStorage
+    :: Foldable f
+    => f (StorableEvent ScriptTxHandle)
+    -> ScriptTxHandle
+    -> IO ScriptTxHandle
+  persistToStorage es h = do
+    let rows = foldl' (\ea e -> ea ++ flatten e) [] es
+        c    = hdlConnection h
+    SQL.execute_ c "BEGIN"
+    forM_ rows $
+      SQL.execute c "INSERT INTO script_transactions (scriptAddress, txCbor, slotNo) VALUES (?, ?, ?)"
+    SQL.execute_ c "COMMIT"
+    pure h
+
+    where
+      flatten :: StorableEvent ScriptTxHandle -> [ScriptTxRow]
+      flatten (ScriptTxEvent txs sl) = do
+        (tx, scriptAddrs) <- txs
+        addr <- scriptAddrs
+        pure $ ScriptTxRow { scriptAddress = addr
+                           , txCbor        = tx
+                           , txSlot        = sl
+                           }
+
+  {- We want to potentially store data in two formats. The first one is similar (if
+     not identical) to the format of data stored in memory; it should contain information
+     that allows knowing at which point the data was generated.
+
+     We use this first format to support rollbacks for disk data. The second format,
+     which is not always necessary and does not have any predetermined structure,
+     should be thought of as an aggregate of the previously produced events.
+
+     For this indexer we don't really need an aggregate, so our "aggregate" has almost the same
+     structure as the in-memory data. We pretend that there is an aggregate by
+     segregating the data into two sections, by using the `hdlDiskStore` parameter. We
```

Comment:
The `hdlDiskStore` doesn't seem to be used anywhere.

Reply:
It is used here: https://github.com/input-output-hk/plutus-apps/pull/825/files#diff-b167f09d5ef11e569397b644012c7dda8a65fd0e7b74dd34bb54b41456eb8eb2R229

Reply:
This is an example of why I don't think that `newtype`-ing over a SQLConnection to make a handle is such a bad thing. I would like to believe that we will need specific configurations for most indexers.